Projeto - 5

AlanDouglasASouza opened the pull request:
Add the permanence of `TODO's` and new options to the user, with a struct 'todos' that makes a `CRUD` allowing the insertion, update, deletion and listing of `TODO's`.  [(Projeto - 5)](https://github.com/terramagna/onboarding-rust/tree/main/essenciais/2-intermediario/03-projeto-5).

AlanDouglasASouza committed:
feat: Add per-session TODO permanence functionality

Add the permanence of TODO's and new options to the user, with a struct 'todos' that makes a CRUD allowing the insertion, update, deletion and listing of TODO's.

lffg commented on src/todos.rs:
```diff
@@ -0,0 +1,58 @@
+    pub fn show_all_todos(&mut self, show_keys: bool) -> Result<(), TerminalError> {
+        let mut terminal = Terminal::new();
+        let todos = &self.all_todos;
+        let mut keys: Vec<&u8> = todos.keys().collect();
+        keys.sort();
```
Se você usar um `BTreeMap`, as chaves retornadas já vêm em ordem!

AlanDouglasASouza replied:
Dei uma boa olhada no BTreeMap. Muito Show.

lffg commented on src/todos.rs:
```diff
@@ -0,0 +1,58 @@
+        keys.sort();
+
+        for key in keys {
+            if let Some(message) = todos.get(key) {
```
Teoricamente, esse `if let Some(_)` aqui não é necessário, concorda? Isso porque você está iterando sobre as _chaves_ que o próprio mapa retornou, logo acima; então você _sabe_ que todas as chaves estarão ali.

Para eliminar esse `if`, uma opção é iterar sobre cada _entry_ do mapa. Uma _entry_ é composta pela chave _e_ pelo valor. Todavia, em `HashMap`s, esses _entries_ seriam retornados em uma ordem qualquer, já que o HashMap não garante nenhuma ordem em relação à iteração.

A `BTreeMap`, entretanto, garante que a iteração (seja pelas chaves, pelas _entries_, etc) sempre irá ocorrer de acordo com a ordem natural das chaves, o que é exatamente o que você precisa aqui. :)

Você pode usar uma referência ao mapa direto no `for`, já que ele é um iterável (i.e., implementa a trait `IntoIterator`)! [Ref.](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#impl-IntoIterator-for-%26%27a%20BTreeMap%3CK%2C%20V%2C%20A%3E)

AlanDouglasASouza replied:
Traz uma vantagem grande mesmo o uso do `BTreeMap` e ali não era necessario o `if let Some(_)`, tinha me passado despercebido.

lffg commented on src/todos.rs:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    pub fn insert_todo(&mut self, todo: Todo) {
+        self.length += 1;
```
Não acha `u8` muito pequeno? Eu usaria pelo menos um `u32`.

AlanDouglasASouza replied:
Concordo, mudei para `u32`.

lffg commented on src/terminal.rs:
```diff
@@ -51,25 +54,35 @@ impl Terminal {
         match self {
             Self::StdoutErr(err) => format!("Houve um erro ao tentar exibir mensagem {}", err),
             Self::StdinErr(err) => format!("Houve um erro na entrada de dados {}", err),
+            Self::ParseErr(err) => format!("Houve um erro na entrada de dados {}", err),
```
Houve um erro ao tentar _analisar_ os seus dados?

AlanDouglasASouza replied:
Tinha esquecido de mudar :P

lffg commented on src/terminal.rs:
```diff
@@ -51,25 +54,35 @@ impl Terminal {
     }
+
+    pub fn ask_which_todo(&mut self) -> Result<u8, TerminalError> {
+        let user_response = self.input.read_line().map_err(TerminalError::StdinErr)?;
```
Essa linha aqui está repetida em vários lugares. O que acha de criar um método `input` (privado) que retorna uma string lida do terminal?

AlanDouglasASouza replied:
Feito!

lffg commented on src/main.rs:
```diff
@@ -14,28 +18,117 @@ fn main() {
 }
 
 fn run() -> Result<(), TerminalError> {
```
Essa função está muito grande. Como você faria para separá-la em blocos menores?

AlanDouglasASouza replied:
Preferi deixar os métodos de interação com o `Terminal`. Isso diminuiu a função para menos da sua metade. WDYT?

lffg commented on src/main.rs:
```diff
@@ -14,28 +18,117 @@ fn main() {
+                                );
+                            }
+                        }
+                        Err(_) => {
+                            terminal.clean()?;
+                            println!(
+                                "❗ {} ❗\n",
+                                red.apply_to("O identificador do TODO precisa ser um número!")
+                            )
+                        }
```
Essa lógica está duplicada no braço de baixo. Por que você não move essa lógica para a parte de report de erros? Isso está muito ligado a sua variante `ParseErr`.

AlanDouglasASouza replied:
Fez mais sentido.

lffg commented on src/main.rs:
```diff
@@ -14,28 +18,117 @@ fn main() {
+            }
+            UserResponse::Update => {
+                terminal.clean()?;
+                loop {
```
O que acontece se o usuário apertar "update" ou "delete" sem ter nenhum todo cadastrado? :P

AlanDouglasASouza replied:
Realmente ficava preso, alterei a lógica para verificar se há algum **TODO** registrado  antes de permitir a entrada no `loop`.

lffg commented on src/main.rs:
```diff
@@ -14,28 +18,117 @@ fn main() {
+                                    green.apply_to("TODO atualizado com sucesso!").bold()
+                                );
+                                break;
+                            } else {
+                                terminal.clean()?;
+                                println!(
+                                    "❗ {} ❗\n",
+                                    red.apply_to("O TODO que você digitou não existe")
+                                );
+                            }
```
Essa lógica também está duplicada.

Faz sentido fazer com que o método `get_one_todo` retorne um `Option` (e não, e.g, um `Result<Todo, SomeError>`), já que isso dá mais _flexibilidade_ ao chamador fazer o que ele quiser se tiver recebido um `None`.

Todavia, ficar repetindo essa tratativa de erro fica meio verboso. O que acha de criar uma nova variante no seu enum de erros `NotFound` e criar um método com esta assinatura:

```rs
fn or_not_found(maybe_todo: Option<&Todo>) -> Result<&Todo, TerminalError> {
    // Converter None na variante `NotFound`
}
```

P.S. Tem um método disponível em `Option` que vai deixar essa implementação bem legal! Dê uma olhada. :)

Aí você pode passar o `Option` que o `get_one_todo` retorna para essa função e usar a `?` depois, deixando a lógica de reportar o erro para outra parte do programa, o que ajuda a evitar a repetição.

```rs
let result = or_not_found(list_todos.get_one_todo(key))?;
// ...
```

---

Essa lógica é bem parecida com o método `or_not_found` que temos na biblioteca `http_problem`. É muito útil pois te permite converter qualquer `Option` em um erro que representa o erro HTTP 404, de maneira extremamente fluente. Nesse caso, você faria:

```rs
let result = list_todos.get_one_todo(key).or_not_found("todo", key)?;
```

Só que como não estamos usando o `problem`, implementamos uma versão mais primitiva na mão. :)

AlanDouglasASouza replied:
Parece muito interessante e útil esse método `or_not_found` do `http_problem`. O método de `Options `que encontrei para essa implementação foi o `ok_or`.

lffg replied:
Sim! O `ok_or` funciona bem aqui também. Bom achado :)

lffg reviewed (commented):
Ficou muito bom! A CLI ficou sensacional!

Mas sempre há espaço para melhorar! :D Vamos ver alguns pontos.

AlanDouglasASouza committed:
chore: modifild methods and run function

AlanDouglasASouza committed:
chore: add use crate

AlanDouglasASouza committed:
refac: Reduce run function and adds collection improvement

Refactoring of the run() function and adding methods in Terminal. Modification of the collection of TODOs for BTreeMap.

AlanDouglasASouza committed:
refac: add implementing the ok_or method on or_not_found

lffg commented on src/terminal.rs:
```diff
@@ -16,60 +19,168 @@ impl Terminal {
+        )?;
+        let key = self.ask_which_todo(list_todos)?;
+        let new_todo = self.ask_for_new_todo()?;
+        list_todos.update(key, new_todo);
```
Ficou bem interessante mover a maior parte da lógica de apresentação para dentro do `Terminal`, mas _essa parte_ especificamente (atualizar a `list_todos`) me parece meio deslocada aqui dentro.

Digo isso porque essa atualização faz parte do _domínio_ da sua aplicação, mas o `Terminal` faz parte da _camada de apresentação_.

O que você acha de fazer com que esse método, `update_todo`, retorne o ID do TODO a ser atualizado, bem como a nova mensagem? Algo como `Result<(u32, String), TerminalError>`. Aí você consegue fazer a atualização _fora_ deste método. Nesse caso, você provavelmente irá concordar comigo que o nome `update_todo` deverá ser alterado para algo como `get_update_todo_data`, ou algo assim.

É claro que depois da atualização você ainda terá que limpar o terminal e mostrar a mensagem de sucesso. Para isso, você pode introduzir um novo método, `write_feedback`, que aceita a string e faz o resto do que tiver que ser feito. :)

O que acha?

AlanDouglasASouza replied:
Interessante, seguindo essa lógica alterei também o método `remove` para a camada de _domínio_.

lffg reviewed (commented):
Ficou ótimo! Só mais um detalhe...

AlanDouglasASouza committed:
refac: moves 'update' and 'remove' from Terminal to domain layer

Brings the responsibility of calling the update and removal of a TODO from the list to the run function. Creation of a feedback method in Terminal.

lffg commented on src/terminal.rs:
```diff
@@ -16,60 +19,155 @@ impl Terminal {
+        eprintln!("{}\n", style(error.message_err()).red().bold());
+    }
+
+    pub fn update_todo(&mut self, list_todos: &mut Todos) -> Result<(u32, Todo), TerminalError> {        
```
Acho que aqui você não precisa mais de uma ref mut.

AlanDouglasASouza replied:
Sim! Na verdade em alguns lugares a mais... Não havia percebido disso. Valeu!

lffg commented on src/terminal.rs:
```diff
@@ -16,60 +19,155 @@ impl Terminal {
+        Ok(key)
+    }
+
+    pub fn check_list_is_empty(&mut self, list: &mut Todos) -> bool {
```
Mesmo de cima.

lffg commented on src/terminal.rs:
```diff
@@ -16,60 +19,155 @@ impl Terminal {
+        Ok((key, new_todo))        
+    }
+
+    pub fn delete_todo(&mut self, list_todos: &mut Todos) -> Result<u32, TerminalError> {        
```
Mesmo de cima.

AlanDouglasASouza replied:
Feito

lffg commented on src/terminal.rs:
```diff
@@ -16,60 +19,155 @@ impl Terminal {
     }
 
     pub fn should_ask_for_todo(&mut self) -> Result<UserResponse, TerminalError> {
```
Agora eu mudaria o nome desse método para `get_user_response` ou algo do tipo. Talvez mudar o nome do método para `get_user_command` e o enum para `UserCommand` para deixar menos genérico, mas isso é nitpick meu.

AlanDouglasASouza replied:
Na verdade eu subi a PR e me esqueci completamente de alterar os nomes :O. Ai eu mudei os nomes desde os métodos `delete_todo` e `update_todo` junto com essa sua sugestão.

lffg reviewed (approved):
Ficou muito bom! :D

Já vou deixar aprovado porque essas mudanças são bem pequenas...

luisholanda commented on src/main.rs:
```diff
@@ -14,28 +18,68 @@ fn main() {
+            UserResponse::Update => {
+                terminal.clean()?;
+
+                loop {
+                    if !terminal.check_list_is_empty(&mut list_todos) {
+                        break;
+                    }
```
Por que não usar um `while` aqui?

```suggestion
                while terminal.check_list_is_empty(&mut list_todos) {
```

AlanDouglasASouza replied:
Show! Não tinha visto essa possibilidade.

luisholanda commented on src/main.rs:
```diff
@@ -14,28 +18,68 @@ fn main() {
+            }
+            UserResponse::Delete => {
+                terminal.clean()?;
+                loop {
+                    if !terminal.check_list_is_empty(&mut list_todos) {
+                        break;
+                    }
```
Mesma coisa aqui.

luisholanda commented on src/terminal.rs:
```diff
@@ -16,60 +19,155 @@ impl Terminal {
+            "\nEscolha uma opção para usar seu TODO LIST 🤔",
+            Style::new().magenta(),
+        )?;
+        self.write_styled("\n1 - Para CRIAR um TODO \n2 - Para LISTAR todos os seus TODO's \n3 - Para ALTERAR um TODO existente \n4 - Para DELETAR um TODO \n5 - Para SAIR\n", Style::new().white())?;
```
Por que não usar uma string multiline aqui?

```suggestion
        self.write_styled(r"
1 - Para CRIAR um TODO
2 - Para LISTAR todos os seus TODO's
3 - Para ALTERAR um TODO existente
4 - Para DELETAR um TODO \n5 - Para SAIR", Style::new().white())?;
```

AlanDouglasASouza replied:
Ficou mais legível. Obrigado!

luisholanda reviewed (approved):
Só uns detalhes, já vou deixar aprovado.

AlanDouglasASouza committed:
refac: remove mutable references from Terminal

Remove mutable references from Terminal and change loop in function run to while.
